### ISSUE TITLE
"Tuple index out of range" false positive despite "len() == 1" check

### ISSUE THREAD

**Issue by dgrunwald** [python/mypy#8733]
```py
import typing 

def test(lookup: typing.Dict[str, typing.List[str]], key: str) -> None:
    vals = lookup.get(key, ())
    # reveal_type(vals) - Revealed type is 'Union[builtins.list[builtins.str], Tuple[]]'
    if len(vals) != 1:
        return
    # reveal_type(vals) - Revealed type is 'Union[builtins.list[builtins.str], Tuple[]]'
    print(vals[0]) # false positive: Tuple index out of range
```

I expect this code to be accepted by the type-checker without errors.
If `len(vals) == 1`, the `Tuple[]` case is impossible. The revealed type should change to just `builtins.list[builtins.str]` after the condition.

Mypy 0.770 already manages this with a `if not vals: return` check; it should also use `len(tuple) <relop> integer-literal` conditions to filter out tuple types that don't match the expected length.

**Comment by msullivan** [comment]
That's a plausible request for a type-narrowing feature that I'd review a pull request for if one was submitted

**Comment by abey79** [comment]
I ran into what I think is a similar issue today:

```python
def demo(param: Union[Tuple[()], Tuple[int, int]]) -> Tuple[int, int]:
    if len(param) == 2:
        return param
    else:
        return 1, 1
```

Despite the `len` check, mypy doesn't recognise that the returned tuple is always of length 2 and yield this error:

```
 error: Incompatible return value type (got "Union[Tuple[], Tuple[int, int]]", expected "Tuple[int, int]")
```

@msullivan let me know if a new issue is warranted for this.

**Comment by ilevkivskyi** [comment]
Duplicate of https://github.com/python/mypy/issues/1178
